draft-ietf-tls-esni.md

```diff
@@ -331,27 +331,39 @@ ignore the `ECHConfig`.
 
 # The "encrypted_client_hello" Extension {#encrypted-client-hello}
 
-The encrypted ClientHelloInner is carried in an "encrypted_client_hello"
-extension, defined as follows:
+To offer ECH, the client sends an "encrypted_client_hello" extension in the
+ClientHelloOuter. When it does, it MUST also send the extension in
+ClientHelloInner.
 
 ~~~
     enum {
        encrypted_client_hello(0xfe0a), (65535)
     } ExtensionType;
 ~~~
 
-When offered by the client, the extension appears only in the ClientHelloOuter.
-The payload MUST have the following structure:
+The payload of the extension has the following structure:
 
 ~~~~
+    enum { is-outer(0), is-inner(1) } ClientECHType;
+
     struct {
-       HpkeSymmetricCipherSuite cipher_suite;
-       uint8 config_id;
-       opaque enc<1..2^16-1>;
-       opaque payload<1..2^16-1>;
+       ClientECHType type;
```

Comment:
Why do we need a type byte? Don't the contents of the extension distinguish inner vs outer? (That is, if it's empty, the extension is of type is-outer, else it's is-inner.) The type byte seems redundant.

Reply:
It avoids an ambiguity between ClientHelloInner with ClientHelloOuterAAD.outer_hello, which contains an empty ECH extension. Plus, using an enum here seemed more idiomatic for TLS.

Reply:
Hmm... why is avoiding that ambiguity important?

Reply:
It might not be. If you prefer to save a byte here, we can do away with the type. I just think the protocol is cleaner and the spec is easier to understand with the type.

Reply:
Yeah, that's fine. I just wanted to make sure I wasn't missing something else. If folks think it's more clear, I'm fine with keeping it.

Reply:
I'm not sure if there is a way to say `select (empty)` in the TLS presentation syntax. Though I suspect the ambiguity doesn't matter. (If it did, we could go back to the idea of just clearing `payload` rather than replacing with an empty extension.)

*shrug* Whatever works. I don't think it's worth worrying about too much.